File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int l = 0;
        int r = matrix.size() - 1;

        int mv = 0;
        while (l <= r) {
            mv = (r - l) + l;

            if (matrix[mv][0] == target) {
                return true;
            } else if (matrix[mv][0] > target) {
                r = mv - 1;
            } else {
                l = mv + 1;
            }
        }

        l = 0;
        r = matrix[0].size() - 1;

        while (l <= r) {
            int m = (r - l) + l;

            if (matrix[mv][m] == target) {
                return true;
            } else if (matrix[mv][m] > target) {
                r = m - 1;
            } else {
                l = m + 1;
            }
        }

        return false;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []
        
        def backtrack(elems: List[int], i: int):
            if sum(elems) == target:
                ans.append(elems.copy())
                return
            if i >= len(nums) or sum(elems) > target:
                return
            
            elems.append(nums[i])
            backtrack(elems, i)
            elems.pop()
            backtrack(elems, i + 1)
            

        backtrack([], 0)

        return ans